[Tang Zhi Lin] iP

Tex-Tang opened the pull request:
# Ted

Ted is a ~~bot~~ (A.I.) that can help you **quickly** manage todo, deadline and event tasks.

### Features
- ⚡️ Instantly fast
- 🛠️ Rich Features 

### How to use?
1. Find the latest release [here](https://github.com/Tex-Tang/ip/releases)
2. Download it and run `java -jar <path/to/jar>`

### What's next?
- [x] Manage multiple types of tasks
- [ ] Reminder (coming soon)
- [ ] Notification (coming soon)

If you are Java Programmer, please consider switch to Python
As quote,
> Any fool can write code that a computer can understand. Good programmers write code that humans can understand.
--Martin Fowler [source](https://www.amazon.com/gp/product/0201485672)

```java
public class Main {
    public static void main(String[] args) {
        // Declaring variables
        int x = 12, y = 10;
        boolean isTrue = true;
        String greeting = "Welcome!";
     }
}
```
```python
x, y = 12, 10
isTrue = True
greeting = "Welcome!"
```
References: https://www.imaginarycloud.com/blog/python-vs-java/

Commit:
Add Gradle support

Tex-Tang committed:
Update Duke class

* add bye command that exit the program
* print out the command if it is a command other than bye

Tex-Tang committed:
Add list and add commands to Duke

Tex-Tang committed:
Refactor Duke class

Duke class: inputs is an array of inputs and a local variable in main method.

inputs array is an array that will be used across multiple commands, refactor it to be class variable can be shared across multiple methods.

* Refactor inputs array to be class variable
* Rename `inputs` array to `tasks`

Tex-Tang committed:
Add mark and unmark functionalities to Duke bot

* Add Task class to encapsulate task's information
* Add mark and unmark commands

Tex-Tang committed:
Add support for different types of task

* Add Todo, Event and Deadline class
* Add `todo`, `event` and `deadline` commands

Tex-Tang committed:
Update testing input and expected output

Tex-Tang committed:
Rename Duke to Ted

Tex-Tang committed:
Add InvalidInputException

Tex-Tang committed:
Handling invalid input at each command

Tex-Tang committed:
Update test input and expected output

Tex-Tang committed:
Add delete function to Ted

Tex-Tang committed:
Add test for delete command

Tex-Tang committed:
Rename methods name

Rename name of command handling method for better readability.

Tex-Tang committed:
Fix code style

Tex-Tang committed:
Add save/load tasks functionalities

Tasks is not persistent.

Save the tasks to files so that we can retrieve in next program run.

* add a Storage class to help us handle file IO operation
* add encode and decode method to Task, ToDo, Deadline and Event classes

Tex-Tang committed:
Add Date parsing and formatting in deadline command

Tex-Tang committed:
Add test for date parsing and formatting

Tex-Tang committed:
Merge branch 'branch-Level-8'

Tex-Tang committed:
Handle InvalidInputException when loading task from file

Tex-Tang committed:
Refactor Storage class

Storage is not re-usable enough, need to specify file path every time.

* define filePath as instance variable so we can re-use it
* make loadTasks and saveTasks non-static

Tex-Tang committed:
Refactor code for better OOP design

Tex-Tang committed:
Update test

Tasks are not empty when exit the program.

Clean up all the tasks once exit.

Tex-Tang committed:
Organize files into packages

Classes are scattered around and it is very messy.

Organize them into different packages.

Tex-Tang committed:
Merge branch 'add-gradle-support'

Tex-Tang committed:
Add grade into project

Tex-Tang committed:
Add unit tests for commands

To ensure program is running expectedly, I added tests to test some of
the command classes.

Tex-Tang committed:
Add .jar related files

Tex-Tang committed:
Remove extra line in initial output

Tex-Tang committed:
Add JavaDoc comments to each file

To make the code more readable and people can understand better.

Tex-Tang committed:
Fix coding style

Tex-Tang committed:
Add Find tasks functionality to Ted

Users may have many tasks, find can help them locate the task quickly.

* Add find command to parser
* Implement find command function by checking task's description contain
  user's search string

Tex-Tang committed:
Rename i to index for better readability

Tex-Tang committed:
Merge branch 'branch-A-CodingStandard'

Tex-Tang committed:
Merge branch 'branch-Level-9'

Tex-Tang committed:
Fix coding style

wrewsama commented on src/main/java/ted/Ui.java:
```diff
@@ -0,0 +1,98 @@
+    private static final String INPUT_PREFIX = "> ";
+
+    /**
+     * Scanner to scan user input
```
Small point but I think this comment is a bit self-explanatory and could be omitted

Tex-Tang replied:
I think it is required in our coding standard, will check it out!

wrewsama commented on src/main/java/ted/task/Deadline.java:
```diff
@@ -0,0 +1,80 @@
+        this.by = by;
+
+        try {
+            if (this.by.toLowerCase().equals("today")) {
```
This feature is really cool

Tex-Tang replied:
Thanks 😄

wrewsama commented on src/main/java/ted/task/TaskList.java:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * Get a task from tasks list
+     * @param index
+     * @return
```
I think you forgot to complete some of the javadocs in this class

Tex-Tang replied:
Good catch!! Will fix it!

wrewsama reviewed (commented):
LGTM!

Tex-Tang committed:
Add CheckStyle configuration files

To maintain code quality and standard across the team

Tex-Tang committed:
Fix code style

Tex-Tang committed:
Add GUI to Ted

We need GUI to provide better user experience.

* Add javafx package
* Add MainWindow
* Add DialogBox to show message of user and Ted

Tex-Tang committed:
Merge branch 'A-CheckStyle'

# Conflicts:
#	src/main/java/ted/Ted.java
#	src/main/java/ted/command/ByeCommand.java
#	src/main/java/ted/command/Command.java
#	src/main/java/ted/command/DeadlineCommand.java
#	src/main/java/ted/command/DeleteCommand.java
#	src/main/java/ted/command/EventCommand.java
#	src/main/java/ted/command/ListCommand.java
#	src/main/java/ted/command/MarkCommand.java
#	src/main/java/ted/command/TodoCommand.java
#	src/main/java/ted/command/UnknownCommand.java
#	src/main/java/ted/command/UnmarkCommand.java

Tex-Tang committed:
Fix incomplete comment

Tex-Tang committed:
Remove unused code

There are several parts of code is unused. Remove that will make
the code more readable and clean.

Remove the unused code.

It will reduce confusion when hand over to other programmers.

Tex-Tang committed:
Add assertions

Assertions can help us to locate programmers' fault easily.

Add some assertions in code

Assertion can ensure programmers will not make careless mistakes
and cause runtime errors when release.

Refer to here: https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/errorHandling.html#assertions

Tex-Tang committed:
Change better parameters' naming

Inconsistent naming will confuse programmers.

Change parameters' naming from `ui` to `uiController`.

Tex-Tang committed:
Add help command

Users do not know all the functionalities provided by our bot. A help
command will list out all the available commands to user so they can use
them.

* add a help command
* show all available commands and usages to user

Tex-Tang committed:
Fix dialog box overflow

Dialog box will cause the text overflow and cannot be seen by user. It
is not user-friendly.

* add a maximum width to label
* wrap label in TextFlow so we can create a dynamic width label box

Tex-Tang committed:
Merge pull request #1 from Tex-Tang/branch-A-Assertions

Add assertions

Tex-Tang committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/ted/Ted.java
#	src/main/java/ted/ui/MainWindow.java

Tex-Tang committed:
Merge pull request #2 from Tex-Tang/branch-A-CodeQuality

Improve code quality

Tex-Tang committed:
Merge pull request #3 from Tex-Tang/branch-C-Help

Add help command

Tex-Tang committed:
Fix code quality issues

Tex-Tang committed:
Improve GUI

The GUI now is not looking modern.

Improve the GUI by adding clean background with dropshadow.

Tex-Tang committed:
Fix comment typo

Tex-Tang committed:
Output useful messages

Tex-Tang committed:
Add Ui photo

Tex-Tang committed:
Update user guide

Tex-Tang committed:
Fix font weight of label

Some system might not have the font.
